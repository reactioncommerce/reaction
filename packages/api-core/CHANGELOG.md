# @reactioncommerce/api-core

## 3.0.0

### Major Changes

- [#6740](https://github.com/reactioncommerce/reaction/pull/6740) [`fee4dbe95`](https://github.com/reactioncommerce/reaction/commit/fee4dbe952e557db8ca658dc08283ba6c7343af9) Thanks [@sujithvn](https://github.com/sujithvn)! - Node18 upgrade includes assert json

### Minor Changes

- [#6684](https://github.com/reactioncommerce/reaction/pull/6684) [`315bb97ab`](https://github.com/reactioncommerce/reaction/commit/315bb97abc3e70dcb1a89da8adca5468302b24be) Thanks [@sujithvn](https://github.com/sujithvn)! - Filter feature. This new feature provides a common function that can be used in a new query endpoint to get filtered results from any collection.

- [#6739](https://github.com/reactioncommerce/reaction/pull/6739) [`0b4b3385f`](https://github.com/reactioncommerce/reaction/commit/0b4b3385f218c4085d8d9f641a28a6e567f4c918) Thanks [@vanpho93](https://github.com/vanpho93)! - feat: upgrade mongodb to verion 5.0

- [#6588](https://github.com/reactioncommerce/reaction/pull/6588) [`9d26f0528`](https://github.com/reactioncommerce/reaction/commit/9d26f05286a43d99f4b8a2924b5a0644a5a03ccc) Thanks [@vanpho93](https://github.com/vanpho93)! - feat: upgrade mongodb to 4.4

### Patch Changes

- Updated dependencies [[`a50a7b359`](https://github.com/reactioncommerce/reaction/commit/a50a7b359bbb546b7abab0e0bfed4c5d8b5ad759), [`315bb97ab`](https://github.com/reactioncommerce/reaction/commit/315bb97abc3e70dcb1a89da8adca5468302b24be), [`fee4dbe95`](https://github.com/reactioncommerce/reaction/commit/fee4dbe952e557db8ca658dc08283ba6c7343af9)]:
  - @reactioncommerce/api-utils@2.0.0
