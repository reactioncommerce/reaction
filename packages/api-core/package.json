{
  "name": "@reactioncommerce/api-core",
  "version": "3.0.0",
  "description": "This NPM package provides the `ReactionAPICore` class. Use this to build a NodeJS microservice that is compatible with the Reaction Commerce platform, or to build your main Reaction Commerce API if you don't want to start by forking the `https://github.com/reactioncommerce/reaction` project.",
  "main": "./src/index.js",
  "type": "module",
  "sideEffects": false,
  "engines": {
    "node": ">=14.18.1"
  },
  "homepage": "https://github.com/reactioncommerce/reaction",
  "url": "https://github.com/reactioncommerce/reaction",
  "email": "engineering@reactioncommerce.com",
  "repository": {
    "type": "git",
    "url": "https://github.com/reactioncommerce/reaction.git",
    "directory": "packages/api-core"
  },
  "author": "Reaction Commerce <hello@reactioncommerce.com>",
  "license": "GPL-3.0",
  "bugs": {
    "url": "https://github.com/reactioncommerce/reaction/issues"
  },
  "scripts": {
    "test": "jest --maxWorkers=4",
    "test:unit:watch": "jest --maxWorkers=4 --watch",
    "test:unit:file": "jest --no-cache --maxWorkers=4",
    "test:unit:file:watch": "jest --no-cache --maxWorkers=4 --watch",
    "lint:gql": "graphql-schema-linter ./src/**/*.graphql"
  },
  "keywords": [],
  "graphql-schema-linter": {
    "rules": [
      "arguments-have-descriptions",
      "deprecations-have-a-reason",
      "descriptions-are-capitalized",
      "enum-values-have-descriptions",
      "enum-values-sorted-alphabetically",
      "fields-have-descriptions",
      "input-object-fields-sorted-alphabetically",
      "input-object-values-have-descriptions",
      "relay-connection-types-spec",
      "relay-page-info-spec",
      "type-fields-sorted-alphabetically",
      "types-are-capitalized",
      "types-have-descriptions"
    ]
  },
  "dependencies": {
    "@apollo/federation": "^0.38.1",
    "@apollo/server": "^4.2.1",
    "@graphql-tools/merge": "^8.3.12",
    "@graphql-tools/schema": "^9.0.10",
    "@reactioncommerce/api-utils": "^2.0.0",
    "@reactioncommerce/logger": "^1.1.3",
    "@reactioncommerce/random": "^1.0.2",
    "@reactioncommerce/reaction-error": "^1.0.1",
    "body-parser": "^1.19.0",
    "callsite": "^1.0.0",
    "cors": "^2.8.5",
    "cuid": "^2.1.8",
    "dataloader": "^2.0.0",
    "diehard": "^1.5.2",
    "dotenv": "^16.0.1",
    "envalid": "^6.0.1",
    "express": "^4.17.1",
    "graphql-iso-date": "^3.6.1",
    "graphql-redis-subscriptions": "^2.5.0",
    "graphql-subscriptions": "^2.0.0",
    "graphql-tag": "^2.10.0",
    "graphql-ws": "^5.11.2",
    "ioredis": "^5.2.4",
    "lodash": "^4.17.15",
    "mongodb": "4.4.1",
    "promise-retry": "^1.1.1",
    "simpl-schema": "^1.7.0",
    "ws": "^8.11.0"
  },
  "devDependencies": {
    "@babel/core": "~7.9.0",
    "@babel/preset-env": "~7.9.0",
    "@reactioncommerce/babel-remove-es-create-require": "~1.0.0",
    "@reactioncommerce/data-factory": "~1.0.1",
    "babel-plugin-rewire-exports": "~2.1.0",
    "babel-plugin-transform-es2015-modules-commonjs": "~6.26.2",
    "babel-plugin-transform-import-meta": "~1.0.1",
    "graphql": "^16.6.0"
  },
  "peerDependencies": {
    "graphql": "^16.6.0"
  },
  "publishConfig": {
    "access": "public"
  }
}
