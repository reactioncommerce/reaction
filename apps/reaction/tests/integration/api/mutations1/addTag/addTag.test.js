import importAsString from "@reactioncommerce/api-utils/importAsString.js";
import insertPrimaryShop from "@reactioncommerce/api-utils/tests/insertPrimaryShop.js";
import encodeOpaqueId from "@reactioncommerce/api-utils/encodeOpaqueId.js";
import Logger from "@reactioncommerce/logger";
import Factory from "/tests/util/factory.js";
import { importPluginsJSONFile, ReactionTestAPICore } from "@reactioncommerce/api-core";

const AddTagMutation = importAsString("./AddTagMutation.graphql");

jest.setTimeout(300000);

let testApp;
let shopId;
let addTag;
let mockTagsAccount;
let tagInput;

beforeAll(async () => {
  testApp = new ReactionTestAPICore();
  const plugins = await importPluginsJSONFile("../../../../../plugins.json", (pluginList) => {
    // Remove the `files` plugin when testing. Avoids lots of errors.
    delete pluginList.files;

    return pluginList;
  });
  await testApp.reactionNodeApp.registerPlugins(plugins);
  await testApp.start();
  shopId = await insertPrimaryShop(testApp.context);

  const adminGroup = Factory.Group.makeOne({
    _id: "adminGroup",
    createdBy: null,
    name: "admin",
    permissions: ["reaction:legacy:tags/create"],
    slug: "admin",
    shopId
  });
  await testApp.collections.Groups.insertOne(adminGroup);

  mockTagsAccount = Factory.Account.makeOne({
    groups: [adminGroup._id],
    shopId
  });
  await testApp.createUserAndAccount(mockTagsAccount);

  addTag = testApp.mutate(AddTagMutation);
});

// There is no need to delete any test data from collections because
// testApp.stop() will drop the entire test database. Each integration
// test file gets its own test database.
afterAll(() => testApp.stop());

beforeEach(async () => {
  tagInput = {
    displayTitle: "Tag: Display Title",
    heroMediaUrl: "mediaurluri",
    isVisible: true,
    metafields: [],
    name: "Tag: Name",
    shopId: encodeOpaqueId("reaction/shop", shopId),
    slug: "Tag: Slug"
  };
});


describe("unauthorized user", () => {
  const accountInternalId = "123";
  let logLevel;
  beforeAll(async () => {
    await testApp.setLoggedInUser({ _id: accountInternalId });
    logLevel = Logger.level();
    Logger.level("FATAL");
  });
  afterAll(async () => {
    await testApp.clearLoggedInUser();
    Logger.level(logLevel);
  });
  test("cannot create tag", async () => {
    expect.assertions(1);
    try {
      await addTag(tagInput);
    } catch (error) {
      expect(error).toMatchSnapshot();
    }
  });
});

describe("authorized user", () => {
  beforeAll(async () => {
    await testApp.setLoggedInUser(mockTagsAccount);
  });
  afterAll(async () => {
    await testApp.clearLoggedInUser();
  });
  afterEach(async () => {
    await testApp.collections.Tags.deleteMany({});
  });
  test("can create tag with slug converted at save", async () => {
    let result;
    try {
      let savedTag = await testApp.collections.Tags.findOne({ slug: "tag-slug" });
      expect(savedTag).toBeNull();
      result = await addTag(tagInput);
      savedTag = await testApp.collections.Tags.findOne({ slug: "tag-slug" });
    } catch (error) {
      expect(error).toBeUndefined();
    }

    // slug is converted when saving slug.
    tagInput.slug = "tag-slug";
    tagInput.heroMediaUrl = `https://shop.fake.site/${tagInput.heroMediaUrl}`;
    delete tagInput.shopId;

    // check that the return tag has the correct information.
    expect(result.addTag.tag).toEqual(expect.objectContaining(tagInput));
  });
  test("can create tag with slug from name at save", async () => {
    let result;
    delete tagInput.slug;
    try {
      let savedTag = await testApp.collections.Tags.findOne({ slug: "tag-name" });
      expect(savedTag).toBeNull();
      result = await addTag(tagInput);
      savedTag = await testApp.collections.Tags.findOne({ slug: "tag-name" });
      expect(savedTag).not.toBeNull();
      expect(savedTag.slug).toEqual("tag-name");
    } catch (error) {
      expect(error).toBeUndefined();
    }
    // slug is converted when saving slug.
    tagInput.heroMediaUrl = `https://shop.fake.site/${tagInput.heroMediaUrl}`;
    delete tagInput.shopId;

    // check that the return tag has the correct information.
    expect(result.addTag.tag).toEqual(expect.objectContaining(tagInput));
  });
  test("can create tag with invalid slug(empty string defaults to name) at save", async () => {
    let result;
    tagInput.slug = "";
    try {
      result = await addTag(tagInput);
    } catch (error) {
      expect(error).toBeUndefined();
    }

    // slug is converted when saving slug.
    tagInput.slug = "tag-name";
    tagInput.heroMediaUrl = `https://shop.fake.site/${tagInput.heroMediaUrl}`;
    delete tagInput.shopId;

    // check that the return tag has the correct information.
    expect(result.addTag.tag).toEqual(expect.objectContaining(tagInput));
  });
  test("cannot create tag with existing slug", async () => {
    const logLevel = Logger.level();
    const duplicateTagSlug = Factory.Tag.makeOne({
      slug: "tag-slug",
      shopId
    });
    await testApp.collections.Tags.insertOne(duplicateTagSlug);
    Logger.level("FATAL");
    expect.assertions(1);
    try {
      await addTag(tagInput);
    } catch (error) {
      expect(error).toMatchSnapshot();
    }
    Logger.level(logLevel);
  });
});
