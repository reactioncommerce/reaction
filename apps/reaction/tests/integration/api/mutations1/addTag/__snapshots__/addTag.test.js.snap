// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`authorized user cannot create tag with existing slug 1`] = `
Array [
  Object {
    "extensions": Object {
      "code": "INTERNAL_SERVER_ERROR",
      "exception": Object {
        "details": Object {},
        "error": "error",
        "eventData": Object {},
        "isClientSafe": true,
        "reason": "Slug tag-slug is already in use",
      },
    },
    "locations": Array [
      Object {
        "column": 3,
        "line": 2,
      },
    ],
    "message": "Slug tag-slug is already in use",
    "path": Array [
      "addTag",
    ],
  },
]
`;

exports[`unauthorized user cannot create tag 1`] = `
Array [
  Object {
    "extensions": Object {
      "code": "FORBIDDEN",
      "exception": Object {
        "details": Object {},
        "error": "access-denied",
        "eventData": Object {},
        "isClientSafe": true,
        "reason": "Access Denied",
      },
    },
    "locations": Array [
      Object {
        "column": 3,
        "line": 2,
      },
    ],
    "message": "Access Denied",
    "path": Array [
      "addTag",
    ],
  },
]
`;
